File: app/src/main/java/wangdaye/com/geometricweather/resource/ResourceHelper.java
package wangdaye.com.geometricweather.resource;

import android.animation.Animator;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.annotation.Size;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.options.NotificationTextColor;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.resource.providers.DefaultResourceProvider;
import wangdaye.com.geometricweather.resource.providers.ResourceProvider;
import wangdaye.com.geometricweather.resource.utils.ResourceUtils;

public class ResourceHelper {

    @NonNull
    public static Drawable getWeatherIcon(ResourceProvider provider,
                                          WeatherCode code, boolean dayTime) {
        return provider.getWeatherIcon(code, dayTime);
    }

    @Size(3)
    public static Drawable[] getWeatherIcons(ResourceProvider provider,
                                             WeatherCode code, boolean dayTime) {
        return provider.getWeatherIcons(code, dayTime);
    }

    @Size(3)
    public static Animator[] getWeatherAnimators(ResourceProvider provider,
                                                 WeatherCode code, boolean dayTime) {
        return provider.getWeatherAnimators(code, dayTime);
    }

    @NonNull
    public static Drawable getWidgetNotificationIcon(ResourceProvider provider,
                                                     WeatherCode code, boolean dayTime,
                                                     boolean minimal, String textColor) {
        if (minimal) {
            switch (textColor) {
                case "light":
                    return provider.getMinimalLightIcon(code, dayTime);

                case "grey":
                    return provider.getMinimalGreyIcon(code, dayTime);

                case "dark":
                    return provider.getMinimalDarkIcon(code, dayTime);
            }
        }

        return provider.getWeatherIcon(code, dayTime);
    }

    @NonNull
    public static Drawable getWidgetNotificationIcon(ResourceProvider provider,
                                                     WeatherCode code, boolean dayTime,
                                                     boolean minimal, boolean darkText) {
        return getWidgetNotificationIcon(
                provider, code, dayTime, minimal, darkText ? "dark" : "light");
    }

    @NonNull
    public static Uri getWidgetNotificationIconUri(ResourceProvider provider,
                                                   WeatherCode code, boolean dayTime,
                                                   boolean minimal, NotificationTextColor textColor) {
        if (minimal) {
            switch (textColor) {
                case LIGHT:
                    return provider.getMinimalLightIconUri(code, dayTime);

                case GREY:
                    return provider.getMinimalGreyIconUri(code, dayTime);

                case DARK:
                    return provider.getMinimalDarkIconUri(code, dayTime);
            }
        }

        return provider.getWeatherIconUri(code, dayTime);
    }

    @NonNull
    public static Drawable getMinimalXmlIcon(ResourceProvider provider,
                                             WeatherCode code, boolean daytime) {
        return provider.getMinimalXmlIcon(code, daytime);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    @NonNull
    public static Icon getMinimalIcon(ResourceProvider provider,
                                      WeatherCode code, boolean daytime) {
        return provider.getMinimalIcon(code, daytime);
    }

    @DrawableRes
    public static int getDefaultMinimalXmlIconId(WeatherCode code, boolean daytime) {
        int id = new DefaultResourceProvider().getMinimalXmlIconId(code, daytime);
        if (id == 0) {
            return R.drawable.weather_clear_day_mini_xml;
        } else {
            return id;
        }
    }

    @NonNull
    public static Drawable getShortcutsIcon(ResourceProvider provider,
                                            WeatherCode code, boolean dayTime) {
        return provider.getShortcutsIcon(code, dayTime);
    }

    @NonNull
    public static Drawable getShortcutsForegroundIcon(ResourceProvider provider,
                                                      WeatherCode code, boolean dayTime) {
        return provider.getShortcutsForegroundIcon(code, dayTime);
    }

    @NonNull
    public static Drawable getSunDrawable(ResourceProvider provider) {
        return provider.getSunDrawable();
    }

    @NonNull
    public static Drawable getMoonDrawable(ResourceProvider provider) {
        return provider.getMoonDrawable();
    }

    @DrawableRes
    public static int getTempIconId(Context context, int temp) {
        StringBuilder builder = new StringBuilder("notif_temp_");
        if (temp < 0) {
            builder.append("neg_");
        }
        builder.append(Math.abs(temp));

        int id = ResourceUtils.getResId(context, builder.toString(), "drawable");
        if (id == 0) {
            return R.drawable.notif_temp_0;
        } else {
            return id;
        }
    }
}

File: app/src/main/java/org/breezyweather/theme/resource/ResourceHelper.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.theme.resource

import android.animation.Animator
import android.content.Context
import android.graphics.drawable.Drawable
import android.graphics.drawable.Icon
import android.net.Uri
import android.os.Build
import androidx.annotation.DrawableRes
import androidx.annotation.RequiresApi
import androidx.annotation.Size
import org.breezyweather.R
import org.breezyweather.common.basic.models.options.NotificationTextColor
import org.breezyweather.common.basic.models.weather.WeatherCode
import org.breezyweather.theme.resource.providers.DefaultResourceProvider
import org.breezyweather.theme.resource.providers.ResourceProvider
import org.breezyweather.theme.resource.utils.ResourceUtils
import kotlin.math.abs

object ResourceHelper {
    fun getWeatherIcon(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean
    ): Drawable {
        return provider.getWeatherIcon(code, dayTime)
    }

    @Size(3)
    fun getWeatherIcons(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean
    ): Array<Drawable?> {
        return provider.getWeatherIcons(code, dayTime)
    }

    @Size(3)
    fun getWeatherAnimators(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean
    ): Array<Animator?> {
        return provider.getWeatherAnimators(code, dayTime)
    }

    fun getWidgetNotificationIcon(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean,
        minimal: Boolean, textColor: String?
    ): Drawable {
        if (minimal) {
            return when (textColor) {
                "light" -> provider.getMinimalLightIcon(code, dayTime)
                "dark" -> provider.getMinimalDarkIcon(code, dayTime)
                else -> provider.getMinimalGreyIcon(code, dayTime)
            }
        }
        return provider.getWeatherIcon(code, dayTime)
    }

    fun getWidgetNotificationIcon(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean,
        minimal: Boolean, darkText: Boolean
    ): Drawable {
        return getWidgetNotificationIcon(
            provider, code, dayTime, minimal, if (darkText) "dark" else "light"
        )
    }

    fun getWidgetNotificationIconUri(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean,
        minimal: Boolean, textColor: NotificationTextColor?
    ): Uri {
        if (minimal) {
            return when (textColor) {
                NotificationTextColor.LIGHT -> provider.getMinimalLightIconUri(code, dayTime)
                NotificationTextColor.DARK -> provider.getMinimalDarkIconUri(code, dayTime)
                else -> provider.getMinimalGreyIconUri(code, dayTime)
            }
        }
        return provider.getWeatherIconUri(code, dayTime)
    }

    fun getMinimalXmlIcon(
        provider: ResourceProvider,
        code: WeatherCode, daytime: Boolean
    ): Drawable {
        return provider.getMinimalXmlIcon(code, daytime)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    fun getMinimalIcon(
        provider: ResourceProvider,
        code: WeatherCode, daytime: Boolean
    ): Icon {
        return provider.getMinimalIcon(code, daytime)
    }

    @DrawableRes
    fun getDefaultMinimalXmlIconId(code: WeatherCode?, daytime: Boolean): Int {
        if (code == null) {
            return R.drawable.weather_clear_day_mini_xml
        }
        val id = DefaultResourceProvider().getMinimalXmlIconId(code, daytime)
        return if (id == 0) {
            R.drawable.weather_clear_day_mini_xml
        } else id
    }

    fun getShortcutsIcon(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean
    ): Drawable {
        return provider.getShortcutsIcon(code, dayTime)
    }

    fun getShortcutsForegroundIcon(
        provider: ResourceProvider,
        code: WeatherCode, dayTime: Boolean
    ): Drawable {
        return provider.getShortcutsForegroundIcon(code, dayTime)
    }

    fun getSunDrawable(provider: ResourceProvider): Drawable {
        return provider.sunDrawable
    }

    fun getMoonDrawable(provider: ResourceProvider): Drawable {
        return provider.moonDrawable
    }

    @DrawableRes
    fun getTempIconId(context: Context, temp: Int): Int {
        val builder = StringBuilder("notif_temp_")
        if (temp < 0) {
            builder.append("neg_")
        }
        builder.append(abs(temp))
        val id = ResourceUtils.getResId(context, builder.toString(), "drawable")
        return if (id == 0) {
            R.drawable.notif_temp_0
        } else id
    }
}
